MDEV-15530: Variable replicate_rewrite_db cannot be found in "show gl…

an3l opened the pull request:
…obal variables"

- Add `replicate_rewrite_db` status variable, that may accept comma
separated key-value pairs.
- Note that option `OPT_REPLICATE_REWRITE_DB` already existed in `mysqld.h`
  from this commit https://github.com/MariaDB/server/commit/23d8586dbfdfdf02fa2f801b9dad91db53025a64

bnestere commented:
Hi Anel!

To provide a general answer to your question from JIRA
> However I don't know how to properly check this new system variable.


I would recommend a test in the rpl MTR suite which ensures that changes to this value on a replica while the server is running are effective. I would first set up the replica with a new database, e.g. 'test2', and an initial rewrite value ('test->test2') and ensure that works as expected by running some transactions on the master and verifying the tables are the same on the primary and replica (e.g. using `source include/diff_tables.inc`). Then, to ensure changes to the new variable are effective at run-time, I would create another new database on the replica, 'test3', reset the slave state (e.g. with the new `source include/reset_slave.inc` to ensure the GTID state is reset as well), change replicate_rewrite_db to `test->test3`, and restart the replica to replay the events from the primary, and ensure the tables are again consistent. Note that I think you have to rebase onto the latest 10.10 branch to get the new file `include/reset_slave.inc`.

I think it would also be okay to extend `rpl/rpl_rewrt_db.test` here.

A few other things to consider:
 1) I think this option should only be allowed to be changed if the replica is not running to ensure that the database can't change in the middle of a transaction. See the variable `opt_slave_parallel_threads` in `sql/sys_vars.cc` for a general idea on how to do this.

 2) Currently, replicate-rewrite-db when specified via the command line will only add new rules (so multiple databases can be rewritten). If we are allowing this value to be changed while the server is running, we should provide a mechanism to reset the value. For example, if a blank value is given, clear all existing rewrite_db rules. Note that this might be necessary for my proposed test, i.e. to change from test->test2 to test->test3.

bnestere commented on sql/sys_vars.cc:
```diff
@@ -5464,15 +5466,19 @@ Sys_var_rpl_filter::global_value_ptr(THD *thd,
        "actual names of table(s) being updated are checked.",
+       PRIV_SET_SYSTEM_GLOBAL_VAR_REPLICATE_REWRITE_DB);
+
+static Sys_var_rpl_filter Sys_replicate_do_db(
```
I think you accidentally switched `Sys_replicate_do_db` with `Sys_replicate_rewrite_db` here

an3l commented:
> Hi Anel!
> 
> To provide a general answer to your question from JIRA
> 
> > However I don't know how to properly check this new system variable.
> 
> I would recommend a test in the rpl MTR suite which ensures that changes to this value on a replica while the server is running are effective. I would first set up the replica with a new database, e.g. 'test2', and an initial rewrite value ('test->test2') and ensure that works as expected by running some transactions on the master and verifying the tables are the same on the primary and replica (e.g. using `source include/diff_tables.inc`). Then, to ensure changes to the new variable are effective at run-time, I would create another new database on the replica, 'test3', reset the slave state (e.g. with the new `source include/reset_slave.inc` to ensure the GTID state is reset as well), change replicate_rewrite_db to `test->test3`, and restart the replica to replay the events from the primary, and ensure the tables are again consistent. Note that I think you have to rebase onto the latest 10.10 branch to get the new file `include/reset_slave.inc`.
> 
> I think it would also be okay to extend `rpl/rpl_rewrt_db.test` here.
> 
> A few other things to consider:
> 
> 1. I think this option should only be allowed to be changed if the replica is not running to ensure that the database can't change in the middle of a transaction. See the variable `opt_slave_parallel_threads` in `sql/sys_vars.cc` for a general idea on how to do this.
> 2. Currently, replicate-rewrite-db when specified via the command line will only add new rules (so multiple databases can be rewritten). If we are allowing this value to be changed while the server is running, we should provide a mechanism to reset the value. For example, if a blank value is given, clear all existing rewrite_db rules. Note that this might be necessary for my proposed test, i.e. to change from test->test2 to test->test3.

Thanks Brandon,
I have addressed all suggestions.
@andrelkin FYI ^

bnestere commented on sql/rpl_filter.cc:
```diff
@@ -749,6 +836,36 @@ Rpl_filter::rewrite_db_is_empty()
+
+  str->length(0);
+
+  while ((s= it++))
```
Two things in this loop to consider:
 1. The separate appends for `-` and `>` can be combined into `append(STRING_WITH_LENGTH("->"))`
 2. Always adding the delimiter can lead to an unnecessary reallocation of the memory to hold the string. If you are avoiding the branch check to see if the string is empty before adding the delimiter, you can initialize a delimiter string and size to "," and 0 for the first iteration, and then set the length to '1' for all other iterations to be both memory and CPU optimized. 
 
 The result would look something like..
```
const char *delimiter= ",";
size_t delim_len= 0;
while ((s= it++))
{
  str->append(delimiter, delim_len); // Will exit early with no change if delim_len is 0
  str->append(s->key, strlen(s->key));
  str->append(STRING_WITH_LENGTH("->"));
  str->append(s->val, strlen(s->val));
  delim_len= 1;
}

// No need for chop()
```

bnestere commented on sql/sql_list.cc:
```diff
@@ -23,13 +23,6 @@
 
 list_node end_of_list;
 
-void free_list(I_List <i_string_pair> *list)
```
Instead of removing `free_list` and creating your own version in `rpl_filter.cc`; you can just add a destructor for the class `i_string_pair` which frees `key` and `val`.

bnestere replied:
Well it was written to be re-used elsewhere (although it never was). If you want to delete it and move it only to where it is used (the `rpl_filter` namespace), I would ask to make it self-contained. I.e., also move the definition of `i_string_pair` into `rpl_filter.h`, change the name `i_string_pair` to no longer imply its use as generic, but something that construes that it is used for rewrite_db logic (e.g. `rewrite_db_rule` ), etc. Even in doing this, I would suggest the destructor pattern and keeping the body of `free_list` as is (although the function name `free_list` should be changed).

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -0,0 +1,103 @@
+--echo # Set replica sys var replicate_rewrite_db
+connection slave;
+--error ER_SLAVE_MUST_STOP
+SET @@GLOBAL.replicate_rewrite_db="test_master->test_replica";
```
Here, the `x->y` rewrite rule should be deleted in favor of only `test_master->test_replica`, right? Can you extend this test to show that updates in database `x` are no longer carried out into `y`? E.g., update `my_table` on the master in `x` and use `include/wait_for_slave_sql_error.inc`

an3l replied:
Done!

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.opt:
```diff
@@ -0,0 +1,2 @@
+"--replicate-rewrite-db=primary_db1->replica_db1"
+"--replicate-rewrite-db=x->y"
```
This option file should only be used by the slave, so it should be renamed `rpl_rewrite_db_sys_vars-slave.opt`. Also a new line should be added to the end of the file.

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -0,0 +1,103 @@
+source include/start_slave.inc;
+
+--source include/rpl_end.inc
+# end of 10.11 tests
```
new line

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrt_db-slave.opt:
```diff
@@ -1 +1,6 @@
+"--replicate-rewrite-db=x->y"
+"--replicate-rewrite-db=database_master_temp_01->database_slave_temp_01"
+"--replicate-rewrite-db=database_master_temp_02->database_slave_temp_02"
+"--replicate-rewrite-db=database_master_temp_03->database_slave_temp_03"
```
This separation into newlines should be added into the cosmetic commit

an3l replied:
Will update this before push, after final review.

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -0,0 +1,103 @@
+--echo # MDEV-15530 Variable replicate_rewrite_db cannot be found
+--echo #            in "show global variables"
+--echo #
+
```
Dynamic updates to the replication filter should be lost after server restart. Can you add a test case to validate that `replicate_rewrite_db` is reverted to the `my.cnf` specification after restart?

bnestere replied:
Ah, yeah `reset_slave.inc` and `start_slave.inc` only work on the slave threads. To restart the mariadbd process, use
```
--let $rpl_server_number= 2
--source include/rpl_restart_server.inc
```

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -0,0 +1,103 @@
+-- source include/have_binlog_format_mixed.inc
+-- source include/master-slave.inc
+
```
Can you write a short comment describing the purpose of the test file here?

bnestere commented on mysql-test/suite/binlog/t/binlog_row_mysqlbinlog_options.test:
```diff
@@ -25,7 +25,7 @@ SET timestamp=1000000000;
 RESET MASTER;
 
-# Whe'll call mysqlbinlog with two rewrite rules:
+# We'll call mysqlbinlog with two rewrite rules:
```
Can you add a new commit for the "cosmetic" changes?

bnestere reviewed (changes requested):
Hi Anel!

Thanks for the patch! Structurally it looks great, I left some suggestions to improve it. I'm happy to answer any questions you may have.

 - Brandon

an3l commented on sql/sys_vars.cc:
```diff
@@ -5505,6 +5511,12 @@ static Sys_var_rpl_filter Sys_replicate_do_db(
 
+static Sys_var_rpl_filter Sys_replicate_rewrite_db(
+       "replicate_rewrite_db", OPT_REPLICATE_REWRITE_DB,
+       "Updates to a database with a different name than the original. Example: "
```
Done.

an3l reviewed (commented):
Hi Brandon, thanks have updated.

bnestere commented on sql/rpl_filter.cc:
```diff
@@ -520,6 +512,70 @@ Rpl_filter::add_wild_table_rule(DYNAMIC_ARRAY* a, const char* table_spec)
 
 
+int
+Rpl_filter::add_string_pair_list(const char* spec)
```
Ok. In that case, I think the name of the function should be refined to be clear that it doesn't follow their convention, and that it specifically targets the `rewrite_db` logic.

bnestere commented on sql/mysqld.cc:
```diff
@@ -8068,45 +8068,7 @@ mysqld_get_one_option(const struct my_option *opt, const char *argument,
-    val= strmake_root(&startup_root, val, (size_t) (ptr-val));
-
-    cur_rpl_filter->add_db_rewrite(key, val);
+    cur_rpl_filter->add_rewrite_db(argument);
```
Ah, the invalid specification of `replicate-rewrite-db=db1` should be done within the `.test` file `mysqlbinlog` invocations rather than the `.opt` file (because `mysqld` won't start otherwise, as you noticed :))

So for example
```
--let $MYSQLD_DATADIR=`select @@datadir`
--let BINLOG_FILENAME= query_get_value(SHOW BINARY LOGS, Log_name, 1)
--let BINLOG_FILE_PARAM= $MYSQLD_DATADIR/$BINLOG_FILENAME

--echo # Invalid specification of replicate-rewrite-db should error
--echo # MYSQL_BINLOG BINLOG_FILE_PARAM --replicate-rewrite-db=db1 2> MYSQLD_VARDIR/tmp/mysqllbinlog.err.out
--error 1
--exec $MYSQL_BINLOG $BINLOG_FILE_PARAM --replicate-rewrite-db=db1 2> $MYSQLD_VARDIR/tmp/mysqlbinlog.err.out

# A quick use of include/search_pattern_in_file.inc or include/assert_grep.inc of mysqlbinlog.err.out to ensure the error message was written
```

an3l replied:
Done in new patch review for `sys_vars.replicate_rewrite_db.test`, using `--exec  $MYSQLD_CMD`, and the result file is updated for `main.mysqlbinlog.test` to reflect the same error message.

bnestere commented on client/mysqlbinlog.cc:
```diff
@@ -2394,7 +2394,11 @@ get_one_option(const struct my_option *opt, const char *argument, const char *fi
-    binlog_filter->add_rewrite_db(argument);
+    if (binlog_filter->add_rewrite_db(argument))
+    {
+      sql_print_error("Bad syntax in replicate-rewrite-db");
```
Maybe add a quick hint that it is expecting FROM->TO (here and `mysqld`)

andrelkin reviewed (approved):
The patch generally looks good, thank you @an3l and @bnestere who made a bunch of a good notes!

bnestere commented on mysql-test/suite/sys_vars/r/replicate_rewrite_db.result:
```diff
@@ -84,5 +84,16 @@ db1->db2,db3->db4
+@@GLOBAL.replicate_rewrite_db
+test->rewrite,mysqltest1->test,a->b,X->Y
+# Check restart with wrong value on CLI
+2022-09-09 11:12:40 0 [ERROR] Bad syntax in replicate-rewrite-db
```
The timestamp here will be problematic on replay. instead of using `cat_file`, I suggest to use `source include/assert_grep.inc`, e.g.

```
--let assert_text= Invalid replicate_rewrite_db CLI argument should produce error
--let assert_count= 1
--let assert_select= [ERROR] Bad syntax in replicate-rewrite-db
--let assert_file=$MYSQL_TMP_DIR/mysqld--replicate.err
--source include/assert_grep.inc
```

an3l replied:
Done with `--replace_regex /mysqld/mariadbd/ /\d\d\d\d-\d*-\d* *\d*:\d*:\d* \d* //`

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -97,6 +101,28 @@ SELECT @@GLOBAL.replicate_rewrite_db;
+--source include/start_slave.inc
+SELECT @@GLOBAL.replicate_rewrite_db;
+--source include/sync_with_master_gtid.inc
+# This should update values with 314 and 1000 from primary (BUT IS NOT !~)
```
Right. The slave's GTID position is already caught up with the master at the time of `rpl_restart_server.inc`. To have the events replayed with the correct filter, reset the slave before the call to `rpl_restart_server.inc`:
```
--source include/stop_slave.inc
--source include/reset_slave.inc
--source include/start_slave.inc
--let $rpl_server_number= $rpl_server_id
--source include/rpl_restart_server.inc
```

bnestere commented on mysql-test/suite/multi_source/replicate_rewrite_db_dynamic.test:
```diff
@@ -0,0 +1,33 @@
+--source include/not_embedded.inc
+--source include/have_innodb.inc
+
+--connect (gandalf_master1,127.0.0.1,root,,,$SERVER_MYPORT_1)
```
For the multisource test, you can reference the `mdev-8874.test` file (in multi_source/) because it sets master-specific filters using multiple servers. A key difference though is that `mdev-8874` uses `mdev-8874.cnf` to set the replication filters, and yours should be specified dynamically (as yours already does :))

an3l replied:
Ok solved

bnestere commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -82,10 +85,11 @@ select * from test_replica.my_table;
 connection master;
 use x;
 insert into my_table values (314);
+select * from my_table;
```
I think it would be good to also show the contents where this is found to be replicated on the slave. I.e., `x.my_table` should have `314` in it.

an3l replied:
Plan is to show that below together with value 1000.

bnestere commented on mysql-test/suite/multi_source/replicate_rewrite_db_dynamic.cnf:
```diff
@@ -0,0 +1,11 @@
+server-id=3
+log-bin=server3-bin
+log-slave-updates
+loose-innodb
```
newline

bnestere reviewed (changes requested):
The patch is coming along nicely! Thanks for the updates, I left just a few more notes.

an3l commented on mysql-test/suite/rpl/t/rpl_rewrite_db_sys_vars.test:
```diff
@@ -0,0 +1,118 @@
+-- source include/have_binlog_format_mixed.inc
```
Have done it in `multi_source` suite, thanks @andrelkin .

bnestere commented on client/mysqlbinlog.cc:
```diff
@@ -2394,48 +2394,11 @@ get_one_option(const struct my_option *opt, const char *argument, const char *fi
+    if (binlog_filter->add_rewrite_db(argument))
     {
-      sql_print_error("Bad syntax in rewrite-db: empty FROM db\n");
+      sql_print_error("Bad syntax in replicate-rewrite-db");
```
Ah, in `mysqlbinlog` the option is just named `rewrite-db` (in `mysqld` it is `replicate-rewrite-db`)

an3l replied:
Done, thanks !

bnestere reviewed (approved):
Approved in [3f909c6](https://github.com/MariaDB/server/commit/3f909c628c1dfac9ae8098c7e4e4d8a81c2e426b). Thank you @an3l for the contribution!

an3l commented on sql/slave.cc:
```diff
@@ -2954,6 +2954,9 @@ void show_master_info_get_fields(THD *thd, List<Item> *field_list,
                         Item_empty_string(thd, "Slave_SQL_Running", 3),
                         mem_root);
+  field_list->push_back(new (mem_root)
+                        Item_empty_string(thd, "Replicate_Rewrite_DB", 23),
```
This  size is for string `Replicate_rewrite_db`, not for value of filter itself. That is done by `protocol::store()`.  The size here is chosen based on longer string, that I have looked into `Replicate_Ignore_Table` that has 23. Not all strings of `field_list` have real size length.

an3l commented:
Note: in order to debug the patch in mtr `replicate_rewrite_db` one should use PR #2244
